test/labeling/test_convergence.py

```diff
@@ -0,0 +1,93 @@
+import unittest
+
+import numpy as np
+import pandas as pd
+import pytest
+
+from snorkel.analysis.utils import set_seed
+from snorkel.labeling.apply.pandas import PandasLFApplier
+from snorkel.labeling.lf import LabelingFunction, labeling_function
+from snorkel.labeling.model.label_model import LabelModel
+from snorkel.labeling.preprocess import preprocessor
+from snorkel.types import DataPoint
+
+
+class LabelingConvergenceTest(unittest.TestCase):
+    @classmethod
+    def setup_class(cls):
+        # Ensure deterministic runs
+        set_seed(123)
+
+        # Create raw data
+        cls.N_TRAIN = 1500
+        cls.N_VALID = 300
+
+        cls.k = 2
+        cls.df_train = create_data(cls.N_TRAIN)
+        cls.df_valid = create_data(cls.N_VALID)
+
+    @pytest.mark.complex
+    def test_labeling_convergence(self) -> None:
+        """Test convergence of end to end labeling pipeline."""
+        # Apply LFs
+        labeling_functions = [f] + [
+            get_labeling_function(divisor) for divisor in range(2, 6)
+        ]
+        applier = PandasLFApplier(labeling_functions)
+        L_train = applier.apply(self.df_train)
+        L_valid = applier.apply(self.df_valid)
+
+        self.assertEqual(L_train.shape, (self.N_TRAIN, len(labeling_functions)))
+        self.assertEqual(L_valid.shape, (self.N_VALID, len(labeling_functions)))
+
+        # Train LabelModel
+        label_model = LabelModel(cardinality=self.k, verbose=False)
+        label_model.train_model(L_train, lr=0.01, l2=0.0, n_epochs=100)
+        Y_lm = label_model.predict_proba(L_train.todense()).argmax(axis=1) + 1
+        Y = self.df_train.y
+        err = np.where(Y != Y_lm, 1, 0).sum() / self.N_TRAIN
+        self.assertLess(err, 0.1)
+
+
+def create_data(n: int) -> pd.DataFrame:
+    X = np.random.random((n, 2)) * 2 - 1
```

Comment:
Quick comment here about what this data set/what the labels represent is would be great.

Reply:
Done.